engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/confluent/serialization/ConfluentKafkaAvroDeserializationSchemaFactory.scala

```diff
@@ -4,36 +4,68 @@ import io.confluent.kafka.serializers._
+class ConfluentKafkaAvroDeserializationSchemaFactory[T: TypeInformation](mode: ConfluentKafkaAvroDeserializationMode,
+                                                                         schemaRegistryClientFactory: ConfluentSchemaRegistryClientFactory,
+                                                                         useSpecificAvroReader: Boolean)
   extends KafkaVersionAwareValueDeserializationSchemaFactory[T] with ConfluentKafkaAvroDeserializerFactory {
 
   override protected def createValueDeserializer(topics: List[String], version: Option[Int], kafkaConfig: KafkaConfig): Deserializer[T] =
-    createDeserializer[T](schemaRegistryClientFactory, kafkaConfig, isKey = false, useSpecificAvroReader)
+    createDeserializer[T](mode, topics.head, version, schemaRegistryClientFactory, kafkaConfig, isKey = false, useSpecificAvroReader)
 }
 
-abstract class ConfluentKeyValueKafkaAvroDeserializationFactory[T: TypeInformation](schemaRegistryClientFactory: ConfluentSchemaRegistryClientFactory, useSpecificAvroReader: Boolean)
+object ConfluentKafkaAvroDeserializationSchemaFactory {
+  def apply[T: TypeInformation](schemaRegistryClientFactory: ConfluentSchemaRegistryClientFactory, useSpecificAvroReader: Boolean): ConfluentKafkaAvroDeserializationSchemaFactory[T] =
+    new ConfluentKafkaAvroDeserializationSchemaFactory(ConfluentKafkaAvroDeserializationMode.STATIC, schemaRegistryClientFactory, useSpecificAvroReader)
+}
+
+abstract class ConfluentKeyValueKafkaAvroDeserializationFactory[T: TypeInformation](mode: ConfluentKafkaAvroDeserializationMode,
+                                                                                    schemaRegistryClientFactory: ConfluentSchemaRegistryClientFactory,
+                                                                                    useSpecificAvroReader: Boolean)
   extends KafkaVersionAwareKeyValueDeserializationSchemaFactory[T] with ConfluentKafkaAvroDeserializerFactory {
 
   override protected def createKeyDeserializer(topics: List[String], version: Option[Int], kafkaConfig: KafkaConfig): Deserializer[K] =
-    createDeserializer[K](schemaRegistryClientFactory, kafkaConfig, isKey = true, useSpecificAvroReader)
+    createDeserializer[K](mode, topics.head, version, schemaRegistryClientFactory, kafkaConfig, isKey = true, useSpecificAvroReader)
 
   override protected def createValueDeserializer(topics: List[String], version: Option[Int], kafkaConfig: KafkaConfig): Deserializer[V] =
-    createDeserializer[V](schemaRegistryClientFactory, kafkaConfig, isKey = false, useSpecificAvroReader)
+    createDeserializer[V](mode, topics.head, version, schemaRegistryClientFactory, kafkaConfig, isKey = false, useSpecificAvroReader)
+}
+
+object ConfluentKafkaAvroDeserializationMode extends Enumeration {
+  type ConfluentKafkaAvroDeserializationMode = Value
+  val NORMAL, STATIC, FRESH = Value
```

Comment:
Let's talk about those modes.
1. Why you have static and fresh? IMO fresh is always OK. Can't see scenario when STATIC will give some value. Which scenario it is?
2. What is normal and opposite to normal for you? :)
What about just two modes: EXPLICIT_VERSION, VERSION_FROM_RECORD? But IMO VERSION_FROM_RECORD should be used only in kafka avro source factory without `version` field. Otherwise it will be misleading

Reply:
You're right, I left: FromSubjectVersion and FromRecord